core/core.go

```diff
@@ -115,54 +140,110 @@ func (c *Core) InsertChain(blocks types.Blocks) (int, error) {
+				hashNumberPriorityList = append(hashNumberPriorityList, hashNumber)
+			}
+			if hashNumber.Number < c.CurrentHeader().NumberU64()+c_maxFutureBlocks {
+				hashNumberList = append(hashNumberList, hashNumber)
+			}
 		}
 	}
+
+	c.serviceBlocks(hashNumberPriorityList)
+	if len(hashNumberPriorityList) > 0 {
+		log.Info("Size of hashNumberPriorityList", "len", len(hashNumberPriorityList), "first entry", hashNumberPriorityList[0].Number, "last entry", hashNumberPriorityList[len(hashNumberPriorityList)-1].Number)
+	}
+	c.serviceBlocks(hashNumberList)
+	if len(hashNumberList) > 0 {
+		log.Info("Size of hashNumberList", "len", len(hashNumberList), "first entry", hashNumberList[0].Number, "last entry", hashNumberList[len(hashNumberList)-1].Number)
+	}
+}
+
+func (c *Core) serviceBlocks(hashNumberList []types.HashAndNumber) {
 	sort.Slice(hashNumberList, func(i, j int) bool {
 		return hashNumberList[i].Number < hashNumberList[j].Number
 	})
 
-	// Only take c_appendQueueThreshold latest blocks out of the database because we know that the
-	// append will be interrupted once we reach the dom block, so no need to get
-	// all the blocks in the appendQueue and load it to the RAM
-	var threshold int
-	if len(hashNumberList) > c_appendQueueThreshold {
-		threshold = c_appendQueueThreshold
-	} else {
-		threshold = len(hashNumberList)
+	var retryThreshold uint64
+	switch common.NodeLocation.Context() {
+	case common.PRIME_CTX:
+		retryThreshold = c_primeRetryThreshold
+	case common.REGION_CTX:
+		retryThreshold = c_regionRetryThreshold
+	case common.ZONE_CTX:
+		retryThreshold = c_zoneRetryThreshold
 	}
+
 	// Attempt to service the sorted list
-	for _, hashAndNumber := range hashNumberList[:threshold] {
-		block := c.GetBlockOrCandidateByHash(hashAndNumber.Hash)
-		if block != nil {
-			c.serviceFutureBlock(block)
+	for _, hashAndNumber := range hashNumberList {
+		header := c.GetHeaderOrCandidateByHash(hashAndNumber.Hash)
+		if header != nil {
+			var numberAndRetryCounter blockNumberAndRetryCounter
+			if value, exist := c.appendQueue.Peek(header.Hash()); exist {
+				numberAndRetryCounter = value.(blockNumberAndRetryCounter)
+				numberAndRetryCounter.retry += 1
+				if numberAndRetryCounter.retry > retryThreshold && numberAndRetryCounter.number+c_appendQueueRemoveThreshold < c.CurrentHeader().NumberU64() {
+					c.appendQueue.Remove(header.Hash())
+				} else {
+					c.appendQueue.Add(header.Hash(), numberAndRetryCounter)
+				}
+			}
+			parentHeader := c.GetHeader(header.ParentHash(), header.NumberU64()-1)
+			if parentHeader != nil {
+				// If parent header is dom, send a signal to dom to request for the block if it doesnt have it
+				_, parentHeaderOrder, err := c.sl.engine.CalcOrder(parentHeader)
+				if err != nil {
+					log.Warn("Error calculating the parent block order in serviceBlocks", "Hash", parentHeader.Hash(), "Number", parentHeader.NumberArray())
+				}
```

Comment:
No return after error, making sure this is intentional

Reply:
good catch, we have to continue in this case